### PR TITLE
Improve flattening with * wildcard

### PR DESCRIPTION
**Pull request by loicleyendecker**
Removed all skips in tests as all functionalities should now be working.

I also rewrote dget to not use recursion as it was easier to make it
work this way and I think it should be more efficient for very big
structures (though probably not a use case).

### CODE REVIEW

**Comment by 1Mark** [comment]
Will review soon 😁

File: deeppath/deeppath.py [hunk 1]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+def _get_repetition_index(key: str) -> Optional[Tuple[str, Union[int, str]]]:
     """Try to match a path for a repetition.
     This will return the key and the repetition index or None if the key
     does not match the expected regex"""
     match = _REPETITION_REGEX.search(key)
     if match:
         key = match.group(1)
         repetition_number = match.group(2)
-        return key, int(repetition_number)
+        if repetition_number != "*":
```
Comment (1Mark) [inline deeppath/deeppath.py:44]: Could we swap the condition? to check for `==` instead please

File: deeppath/deeppath.py [hunk 2]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
-                    data = _flatdget(data, index)
-                elif isinstance(data, Dict):
-                    data = [_flatdget(value, index) for value in data.values()]
-            else:
-                if key != "*":
-                    data = _flatdget(data, key)
-                elif not isinstance(data, Sequence):
-                    data = list(data.values())
-    except (KeyError, TypeError, IndexError):
+def _get_sequence_index(path: str) -> Tuple[bool, Union[None, int, str]]:
```
Comment (1Mark) [inline deeppath/deeppath.py:51]: docstring pls

File: deeppath/deeppath.py [hunk 3]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+    match = re.match(r"\[(-?[\d\*]+)\]", path)
+    if match:
+        index = match.group(1)
+        if index.isnumeric() or index[0] == "-" and index[1:].isnumeric():
+            return True, int(index)
+        else:
+            return True, index
+    if path == "*":
+        return False, "*"
+    return False, None
```
Comment (1Mark) [inline deeppath/deeppath.py:61]: We **could** shorten this to `return False, "*" if path == "*" else None`

File: deeppath/deeppath.py [hunk 4]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
-                    data = list(data.values())
-    except (KeyError, TypeError, IndexError):
+def _get_sequence_index(path: str) -> Tuple[bool, Union[None, int, str]]:
+    match = re.match(r"\[(-?[\d\*]+)\]", path)
+    if match:
+        index = match.group(1)
+        if index.isnumeric() or index[0] == "-" and index[1:].isnumeric():
+            return True, int(index)
+        else:
+            return True, index
```
Comment (1Mark) [inline deeppath/deeppath.py:58]: **Could** do
    
    cast_index_to_int = index.isnumeric() or index[0] == "-" and index[1:].isnumeric()
    return True, int(index) if cast_index_to_int else index

File: deeppath/deeppath.py [hunk 5]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+    if path == "*":
+        return False, "*"
+    return False, None
+
+
+def dget(
+    data: Dict[str, Any], path: str, default: Optional[Any] = None
```
Comment (1Mark) [inline deeppath/deeppath.py:65]:
```suggestion
    data: Mapping, path: str, default: Optional[Any] = None
```
Type hint changed from before, is this intentional?

File: deeppath/deeppath.py [hunk 6]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+        next_path, *remainder = remainder
+        if not next_path:
+            continue
+
+        sequence, index = _get_sequence_index(next_path)
+        if index == "*":
+            repetition_flag = True
+            if not sequence and isinstance(node, Mapping):
+                for value in node.values():
+                    nodes.append((value, remainder))
```
Comment (1Mark) [inline deeppath/deeppath.py:85]:
```suggestion
                    nodes.extend((value, remainder) for value in node.values())
```

File: deeppath/deeppath.py [hunk 7]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+            elif sequence and isinstance(node, Sequence):
+                nodes.extend((val, remainder) for val in node)
+        else:
+            if index is None:
+                try:
+                    nodes.append((node[next_path], remainder))
+                except (KeyError, TypeError):
```
Comment (1Mark) [inline deeppath/deeppath.py:92]: `contextlib.supress`

File: deeppath/deeppath.py [hunk 8]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+                    pass
+            elif isinstance(index, int) and isinstance(node, Sequence):
+                try:
+                    nodes.append((node[index], remainder))
+                except IndexError:
```
Comment (1Mark) [inline deeppath/deeppath.py:97]: `contextlib.supress`

File: tests/test_deeppath.py [hunk 9]
```diff
@@ -126,17 +120,17 @@ def test_dget_flatten_and_repetition():
     assert dget(data1, "*") == [reps]
     assert dget(data1, "flattened[0]") == {"nested_in_rep": 1}
     # */ is a list, it needs explicit unfold
-    assert dget(data1, "*/nested_in_rep") is None
+    assert dget(data1, "*/nested_in_rep") == []
     assert dget(data1, "*[*]/nested_in_rep") == [1, 2]
```
Comment (1Mark) [inline tests/test_deeppath.py:124]: I think there were some edge cases where this kind of behaviour didn't work, please check in Jira issues I did report it. I guess it's different now though since we now have wildcard support.

**Review by 1Mark — CHANGES_REQUESTED** [review]
Looks really good Loic 👍  some minor comments. Please update the changelog.rst.

Optional but would be nice, let's get some examples https://pypi.org/project/deeppath/ which would encourage users to install 🚀 📈

File: deeppath/deeppath.py [hunk 10]
```diff
@@ -32,51 +33,78 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+                    pass
+
+    # If there was an explicit repetition in the path (a "*"), then we return a
+    # list, otherwise, we return a single element
+    if repetition_flag:
+        return output
+    elif output:
+        return output[0]
+    else:
```
Comment (1Mark) [inline deeppath/deeppath.py:106]: is mypy forcing else?

File: tests/test_deeppath.py (hunk 9 above)
Comment (1Mark) [reply tests/test_deeppath.py:124]: Please do check it may or may not have been closed by me in Jira but I think we should be wary if it's going to cause existing versions to break just search for `dget` in the Jira project that tracked issues for this.

File: deeppath/deeppath.py [hunk 11]
```diff
@@ -32,51 +34,93 @@ def flatten(nested_iterable: Iterable[Any]) -> List[Any]:
...
+        index_is_numberic = index.isnumeric() or index[0] == "-" and index[1:].isnumeric()
+        return True, (int(index) if index_is_numberic else index)
+    return False, ("*" if path == "*" else None)
+
+
+def dget(
+    data: Mapping[str, Any], path: str, default: Optional[Any] = None
+) -> Union[List[Any], Any]:
+    """
+    Match a path in a deep container.
```
Comment (1Mark) [inline deeppath/deeppath.py:78]: XPath-like syntax for accessing nested mappings.

**Comment by 1Mark** [comment]
- [ ] Please update the changelog.rst.

**Comment by drusu55** [comment]
Heya, when is this going to be merged, and are you planning a new release after? Need deeppath's help and I'd rather have these changes as well hehe :D

**Comment by loicleyendecker** [comment]
> Heya, when is this going to be merged, and are you planning a new release after? Need deeppath's help and I'd rather have these changes as well hehe :D
@drusu55 I'd like to have this merged in the next week or so at worse, but I have some other (cool, in my opinion) changes lined up, so it might be a big release.

File: tests/test_deeppath.py (hunk 9 above)
Comment (1Mark) [reply tests/test_deeppath.py:124]: It looked correct, I was just worried things may break so I wanted to sanity check it. I can't remember if I had added an xfail test for this, I don't think I did... Ah well.

**Comment by loicleyendecker** [comment]
Fyi, I tested with the added test, and I think the issue was just the lists were not properly handled. I get the point that there is still a  conceptual problem when you expect maybe 3 entries but get only two (which maybe we could address with some stuff like {2, 3} to say you want at least 2 or up to 3 results, but also eww), but I think it is quite natural to just say that you want anything that matches with no regard for the actual number of matches.

**Comment by 1Mark** [comment]
I think we should be ok to merge once
- #17 is merged
- Existing open comments are responded to. 

Thanks Loic :)

**Comment by 1Mark** [comment]
only this is pending https://github.com/loicleyendecker/deeppath/pull/8#issuecomment-1059524180 do you want to do this later ?

**Comment by loicleyendecker** [comment]
> only this is pending [#8 (comment)](https://github.com/loicleyendecker/deeppath/pull

**Comment by loicleyendecker** [comment]
> only this is pending [#8 (comment)](https://github.com/loicleyendecker/deeppath/pull/8#issuecomment-1059524180) do you want to do this later, if so i will merge now?

Actually I think the changelog is updated on release automatically.
